librbd: serialize maintenance operations by type

trociny opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

trociny commented:
@dillaman At this stage it is almost not tested, and actually I expect there may have be issues on releasing lock/shutdown if there are queued operations. I am  writing tests for these scenarios. Right now I am interested to learn if you like the approach in general or have any suggestions.

There are also several cases in when we call `operations->execute_XYZ` internally, which currently bypass the queue (e.g. in deep_copy). I am not sure about those if we want to serialize those too.

dillaman commented on src/librbd/ImageWatcher.cc:
```diff
@@ -1135,14 +1172,28 @@ bool ImageWatcher<I>::handle_payload(const SnapRenamePayload &payload,
-                                                  payload.snap_name,
-                                                  new C_ResponseMessage(ack_ctx));
-      return false;
+      bool new_request;
```
Nit: it would be great if we could get a helper function to handle all of this boilerplate code

dillaman commented on src/librbd/Operations.h:
```diff
@@ -121,6 +126,10 @@ class Operations {
   std::atomic<uint64_t> m_async_request_seq;
 
+  mutable ceph::mutex m_queue_lock;
+  std::set<std::string> m_in_flight_ops;
```
Nit: I'd vote for an enum instead of a free-form string for the key

dillaman commented on src/librbd/Operations.cc:
```diff
@@ -333,7 +346,48 @@ bool needs_invalidate(I& image_ctx, uint64_t object_no,
+
+          auto queued_ctx = queue.front();
+          queue.pop_front();
+          queued_ctx->complete(r);
```
Nit: should you drop the lock before invoking?

trociny replied:
The context is wrapped with `util::create_async_context_callback` so I think it is not necessary to drop the lock.

trociny commented on src/librbd/WatchNotifyTypes.cc:
```diff
@@ -185,17 +186,22 @@ void SnapPayloadBase::decode(__u8 version, bufferlist::const_iterator &iter) {
   if (version >= 6) {
     decode(snap_namespace, iter);
   }
+  if (version >= 7) {
```
The version 7 was introduce when adding quiesce callbacks (63e143344). It is still in the master only. So I thought we would not need to bump it if the PR had been merged before pacific was branched?

trociny commented on src/librbd/WatchNotifyTypes.cc:
```diff
@@ -206,15 +212,11 @@ void SnapCreatePayload::decode(__u8 version, bufferlist::const_iterator &iter) {
     decode(snap_namespace, iter);
   }
   if (version >= 7) {
-    decode(async_request_id, iter);
```
I thought I could break  backwards compatibility here because this code is only in the master.

dillaman replied:
Yes, if version 7 was only on master we are good.

trociny commented:
@dillaman It is still in progress and needs cleanup (enum instead of string, a helper function to avoid code duplicate, as you suggested), but right now I have a question I would like to discuss.

Actually it is illustrated by third "workaround" commit. I needs this to make may "shutdown" test work.

When I have some requests in the queue, and the exclusive lock shutdown is started, it is ok for an operation in progress, as it   has `AsyncRequest` "protection". But when this operation is complete and we should to decide if it is safe to start the next operation in the queue or should complete it with `-ERESTART`, it is not enough just to check for `is_lock_owner`, because it may return true when releasing the lock is already in progress. So as a workaround (just to illustrate the point) I added a new check. 

I suspect it is still racy, and actually I have suspicion than in the current master we also have a small window in `C_InvokeAsyncRequest`, between we make sure the lock is acquired and the request is actually started, when we have a chance that the process of releasing the lock is started before the request (`AsyncRequest`) adds itself to the `async_requests` lists. 

So I just thinking, couldn't we add the `acquire_lock` step in `AsyncRequest`, similarly to what we currently have in `C_InvokeAsyncRequest` and complete the requests with `-ERESTART` if it fails?

trociny commented:
And another thing I forgot to mention. I had to change `operation->queue()` to `operation->start_op()` and `operation->finish_op()` after I realized that as the op requests is async the queue needs a way to detect when a request is finished and another one can be started.

trociny commented:
@dillaman I think I addressed all your comments and the only thing remains is that issue with detecting the exclusive lock transient state when dequeuing an operation.

dillaman commented:
> @dillaman I think I addressed all your comments and the only thing remains is that issue with detecting the exclusive lock transient state when dequeuing an operation.

Could you re-use the existing `ExclusiveLock::start_op` call that `rbd-mirror` uses to prevent the release of the lock?

dillaman commented on src/librbd/WatchNotifyTypes.cc:
```diff
@@ -232,37 +232,51 @@ void SnapRenamePayload::decode(__u8 version, bufferlist::const_iterator &iter) {
 }
 
 void RenamePayload::dump(Formatter *f) const {
+  f->open_object_section("async_request_id");
```
Nit: `syncRequestPayloadBase::dump(f);`

dillaman commented on src/librbd/WatchNotifyTypes.cc:
```diff
@@ -232,37 +232,51 @@ void SnapRenamePayload::decode(__u8 version, bufferlist::const_iterator &iter) {
 }
 
 void UpdateFeaturesPayload::dump(Formatter *f) const {
+  f->open_object_section("async_request_id");
```
Nit: `syncRequestPayloadBase::dump(f);`

dillaman commented on src/librbd/WatchNotifyTypes.cc:
```diff
@@ -288,15 +302,22 @@ void MetadataUpdatePayload::encode(bufferlist &bl) const {
 }
 
 void MetadataUpdatePayload::dump(Formatter *f) const {
+  f->open_object_section("async_request_id");
```
Nit: `syncRequestPayloadBase::dump(f);`

trociny commented:
> Could you re-use the existing ExclusiveLock::start_op call that rbd-mirror uses to prevent the release of the lock? You could complete the returned context whenever it's safe for the lock to be released because the operation has been started and recorded in the in-flight list or has completed.

Indeed, I forgot about `ExclusiveLock::start_op`. I will try. Thanks!

trociny commented:
@dillaman Making all operations async revealed an interesting issue I am not sure how to address.

Now `TestInternal.WriteFullCopyup` can sporadically get stuck, and here is an example with the debug rbd turned on [1].

The problem is that the watcher `handle`, which is just the address of `LingerOp` object [2], may be reused when the image is reopen. And a remote operation request may be send with `async_request_id` that was used on the first open. And the lock owner treats this request as a duplicate.

Here are details for the case [1]. The test opens an image and keeps it opened during the test, being the lock owner. And during the test it creates and protects a snapshot, opening temporarily for this a new context, and the operations are proxied to the lock owner:
```
2020-11-19T23:35:54.162+0000 7f5d77fff700 10 librbd::ImageWatcher: 0x7f5d8c027400 async request: [4581,140039623365024,2]
...
2020-11-19T23:35:53.140+0000 7f5d777fe700 10 librbd::ImageWatcher: 0x7f5d8c00c8b0 remote snap_create request: [4581,140039623365024,1] [user] snap1 0
2020-11-19T23:35:53.140+0000 7f5d777fe700 20 librbd::ExclusiveLock: 0x7f5d24001710 accept_request=1 (request_type=0)
2020-11-19T23:35:53.140+0000 7f5d777fe700 20 librbd::Operations: start_op: snap create 0x7f5d84022200
...

2020-11-19T23:35:54.162+0000 7f5d77fff700 10 librbd::ImageWatcher: 0x7f5d8c027400 async request: [4581,140039623365024,2]
...
2020-11-19T23:35:54.170+0000 7f5d777fe700 10 librbd::ImageWatcher: 0x7f5d8c00c8b0 remote snap_protect request: snap1
2020-11-19T23:35:54.170+0000 7f5d777fe700 20 librbd::ExclusiveLock: 0x7f5d24001710 accept_request=1 (request_type=0)
2020-11-19T23:35:54.170+0000 7f5d777fe700 20 librbd::Operations: start_op: snap protect 0x7f5d8c020a20
```
Then at tear down, the test unprotects and removes all tracked snapshots (the image is still keept open by the lock owner at this time). For this it opens a new context and the snap_unprotect request is proxied:
```
2020-11-19T23:35:57.409+0000 7f5d777fe700 10 librbd::ImageWatcher: 0x7f5d8c01fba0 async request: [4581,140039623365024,1]
...
2020-11-19T23:35:57.415+0000 7f5d77fff700 10 librbd::ImageWatcher: 0x7f5d8c00c8b0 remote snap_unprotect request: snap1
2020-11-19T23:35:57.415+0000 7f5d77fff700 20 librbd::ExclusiveLock: 0x7f5d24001710 accept_request=1 (request_type=0)
2020-11-19T23:35:57.415+0000 7f5d77fff700 10 librbd::Watcher::C_NotifyAck 0x7f5d8805ae70 finish: r=0
```
Note, although it is a new watcher (the `ImageWatcher` has a different address 0x7f5d8c01fba0) it has the same `handle` (140039623365024), and the request is treated by the lock owner as a duplicate.

It was a little surprise to me to learn that a client generates the watcher `handle` (`cookie`) itself. I was always thinking it came from the osd when watch was registering...

[1] http://qa-proxy.ceph.com/teuthology/trociny-2020-11-19_18:32:08-rbd-wip-mgolub-testing-distro-basic-smithi/5637954/teuthology.log
[2] https://github.com/ceph/ceph/blob/master/src/osdc/Objecter.h#L2316

dillaman commented:
> It was a little surprise to me to learn that a client generates the watcher `handle` (`cookie`) itself. I was always thinking it came from the osd when watch was registering...

I think you need to combine the `gid` with the `cookie` to get the true unique id (which is `librbd::watcher::ClientId`). Since the `gid` is re-used when using the same connection, obviously that doesn't help here. Perhaps the simplest fix is to change the `Objecter` to use a sequence for the cookie and have a mapping table from cookie to `LingerOp`?

trociny commented:
> I think you need to combine the gid with the cookie to get the true unique id (which is librbd::watcher::ClientId). Since the gid is re-used when using the same connection, obviously that doesn't help here. Perhaps the simplest fix is to change the Objecter to use a sequence for the cookie and have a mapping table from cookie to LingerOp?

As it looks like it is a librbd only problem I am not very happy complicating the Objecter. If the librbd had some (global or per client) source of unique ids it could use it instead of `Operations::m_async_request_seq`...

trociny commented:
@dillaman do you think there might be issues with the approach in the last commit?

dillaman commented on src/librbd/Operations.h:
```diff
@@ -43,11 +43,10 @@ enum Operation {
 class Operations {
 public:
-  Operations(ImageCtxT &image_ctx);
+  static uint64_t reserve_async_request_id();
```
Nit: unused

dillaman commented on src/librbd/Utils.cc:
```diff
@@ -194,5 +194,11 @@ SnapContext get_snap_context(
 }
 
+uint64_t reserve_async_request_id() {
+  static std::atomic<uint64_t> async_request_seq;
```
Nit: does this correctly initialize the sequence to zero? [1] seems to hint that the wrapped object is not initialized.

[1] https://en.cppreference.com/w/cpp/atomic/atomic/atomic

trociny replied:
I used from C times that all static variables are always initialized to 0 if not explicitly initialized. Ok, I will add the explicit initialization so it does not rise the question.

trociny commented:
@dillaman thanks. updated.

Jenkins failed on `TestMigration.StressLive` but I suppose it is the old known issue [1].

[1] https://tracker.ceph.com/issues/45694

dillaman commented:
jenkins test make check
